geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala

```diff
@@ -10,169 +10,40 @@ package org.locationtech.geomesa.features
-import org.geotools.feature.{AttributeImpl, GeometryAttributeImpl}
-import org.geotools.filter.identity.FeatureIdImpl
-import org.geotools.geometry.jts.ReferencedEnvelope
-import org.geotools.util.Converters
-import org.opengis.feature.`type`.{AttributeDescriptor, Name}
+import org.locationtech.geomesa.features.AbstractSimpleFeature.{AbstractImmutableSimpleFeature, AbstractMutableSimpleFeature}
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
-import org.opengis.feature.{GeometryAttribute, Property}
-import org.opengis.filter.identity.FeatureId
-import org.opengis.geometry.BoundingBox
-
-import scala.collection.JavaConversions._
 
 /**
- * Simple feature implementation optimized to instantiate from serialization
- *
- * @param initialId simple feature id
- * @param sft simple feature type
- * @param initialValues if provided, must already be converted into the appropriate types
- */
-class ScalaSimpleFeature(initialId: String,
-                         sft: SimpleFeatureType,
+  * Simple feature implementation optimized to instantiate from serialization
+  *
+  * @param sft simple feature type
+  * @param initialId simple feature id
+  * @param initialValues if provided, must already be converted into the appropriate types
+  * @param initialUserData user data
+  */
+class ScalaSimpleFeature(sft: SimpleFeatureType,
+                         initialId: String,
                          initialValues: Array[AnyRef] = null,
                          initialUserData: java.util.Map[AnyRef, AnyRef] = null)
-    extends SimpleFeature {
-
-  val featureId = new FeatureIdImpl(initialId)
-  val values = if (initialValues == null) Array.ofDim[AnyRef](sft.getAttributeCount) else initialValues
-
-  lazy private[this] val geomDesc  = sft.getGeometryDescriptor
-  lazy private[this] val geomIndex = if (geomDesc == null) -1 else sft.indexOf(geomDesc.getLocalName)
-  lazy private[this] val userData  =
-    if (initialUserData == null) {
-      new java.util.HashMap[AnyRef, AnyRef]()
-    } else {
-      new java.util.HashMap[AnyRef, AnyRef](initialUserData)
-    }
-
-  def setId(id: String): Unit = featureId.setID(id)
-
-  override def getFeatureType: SimpleFeatureType = sft
-  override def getType: SimpleFeatureType = sft
-  override def getIdentifier: FeatureId = featureId
-  override def getID: String = featureId.getID // this needs to reference the featureId, as it can be updated
-  override def getName: Name = sft.getName
-  override def getUserData: java.util.Map[AnyRef, AnyRef] = userData
+    extends AbstractMutableSimpleFeature(sft, initialId, initialUserData) {
 
-  override def getAttribute(name: Name): AnyRef = getAttribute(name.getLocalPart)
-  override def getAttribute(name: String): AnyRef = {
-    val index = sft.indexOf(name)
-    if (index == -1) null else getAttribute(index)
+  @deprecated("use primary constructor")
+  def this(initialId: String, sft: SimpleFeatureType) = {
```

Comment:
Ah, nice!  I was worried that all kinds of client code was going to have to change.